Add "linear" animation style to "approach different" mod

mk56-spn opened the pull request:
Hello

This is a small PR aiming to fix a few things that have been bugging me whilst playtesting the approach different mod. This is not a comprehensive list of all the issues but it adresses the two most important ones.

**Changed the minimum raidus from 2 to 1.5**

this change aims to make approach different (  with the gravity option selected ) a better companion mod to traceable, with this change the smaller initial radius allows more accurate pinpointing of the circles location, before the gravity approach circle expands and proceeds to function like an approach circle.

Before(2x min):
![osu_2021-11-25_10-34-03](https://user-images.githubusercontent.com/74463310/143416236-38a88db6-435e-4b9f-8d86-c2db8e9492a2.jpg)

After(1.5x min):
![osu_2021-11-25_10-33-51](https://user-images.githubusercontent.com/74463310/143416251-2681d2c1-a9e5-4ef1-830e-c5a03033b840.jpg)

**Implemented linear approach circle**
Approach different was lacking the option to simply change the initial approach circle size without giving it any acceleration/ decceleration. This is now the default option under the name linear. Overall this is just a neat little quality of life change that allows more flexibility in how a player is displayed the approach circles.


This is by no mean the only issues approach different faces, like obscure naming (i had to go into the code to figure out what each different acceleration does and even there its not clear you need to go to another page which details it : https://easings.net/#easeOutCirc ) which i have not fixed since i am not positive about what naming scheme to apply.

mk56-spn committed:
Expand the functionality of Approach different

PercyDan54 commented on osu.Game.Rulesets.Osu/Mods/OsuModApproachDifferent.cs:
```diff
@@ -86,6 +89,7 @@ private Easing getEasing(AnimationStyle style)
 
         public enum AnimationStyle
         {
+            Linear,
```
I feel like this should be "None" because I don't think "Linear" is friendly to user

mk56-spn replied:
you may be right, ill wait for a second opinion and change it if your suggestion is favoured

bdach replied:
Disagree. The related setting is described as the "animation style of the approach circles". Therefore it makes no sense to put "none" there as an option because that just reads as if there was no animations at all, and what would that mean? No approach circles, or just static ones? So I would say "none" is actively misleading in this context.

mk56-spn replied:
im thinking "classic" or "standard"  might be good options, thoughts?

bdach replied:
I'm fine with keeping linear. Standard might also be ok, not sure.

bdach commented:
To me at least one thing needs changing, and that is: now that the "linear" easing is in (and it is selected by default, because it was added first to the enum), and the default initial size setting is also the default 4.0, now if you just put Approach Different on and change no settings then nothing visible changes in the gameplay which is bad.

You could do what Difficulty Adjust does, namely override this:

https://github.com/ppy/osu/blob/7a3c69544b914d33c073c687df9158450a00672a/osu.Game/Rulesets/Mods/ModDifficultyAdjust.cs#L27

Overriding the above to `true` ensures that the mod settings are displayed immediately to the user when they choose the mod, rather than requiring them to know that they have to click the "customise" button for the mod to have any observable effect.

I was never really a huge fan of this mod (always felt like something that was easier to implement rather than something that is actually interesting to play), so I won't comment on the adjustments themselves too much. Code seems sane at the least.

bdach reviewed (changes requested):
see above

mk56-spn commented:
> To me at least one thing needs changing, and that is: now that the "linear" easing is in (and it is selected by default, because it was added first to the enum), and the default initial size setting is also the default 4.0, now if you just put Approach Different on and change no settings then nothing visible changes in the gameplay which is bad.
> 
> You could do what Difficulty Adjust does, namely override this:
> 
> https://github.com/ppy/osu/blob/7a3c69544b914d33c073c687df9158450a00672a/osu.Game/Rulesets/Mods/ModDifficultyAdjust.cs#L27
> 
> Overriding the above to `true` ensures that the mod settings are displayed immediately to the user when they choose the mod, rather than requiring them to know that they have to click the "customise" button for the mod to have any observable effect.
> 
> I was never really a huge fan of this mod (always felt like something that was easier to implement rather than something that is actually interesting to play), so I won't comment on the adjustments themselves too much. Code seems sane at the least.

alrighty will fix this, i was split between leaving default up top or not because of this but your suggestion seems good, could put it at the bottom but that irks me. 

as for your distaste for the mod, i cant say i agree i find it plenty fun, especially gravity and linear, the others not so much

peppy commented:
The default should not be linear. The defaults should not be changed.

mk56-spn committed:
Linear approach type moved per peppys request

mk56-spn commented:
whilst i disagree with not putting linear at the top ( as the closest to normal approach i feel like it only makes sense for this to be the case) i have changed it

peppy commented:
No one said it shouldn't be at the top. I said it shouldn't be the default.

mk56-spn commented:
> No one said it shouldn't be at the top. I said it shouldn't be the default.

alright  ill set gravity to the default

Commit:
Moved linear to a better place.

i decided to go with leaving it under gravity, its plenty visible and fine there, since the public enum list self orders I wasn't sure about how i could set a default that wasn't the topmost option

mk56-spn commented:
as far as i know im done here, but if someone is kind enough to explain how to override the public enum listing order ill change the default

bdach commented:
I don't have a strong opinion as to what the order of the enum should be so I won't push any ordering changes myself. But generally you can use any order of the enum and just change the default by passing it to the ctor of the appropriate bindable like so:

```csharp
        [SettingSource("Style", "Change the animation style of the approach circles.", 1)]
        public Bindable<AnimationStyle> Style { get; } = new Bindable<AnimationStyle>(AnimationStyle.Gravity);
```

mk56-spn committed:
Added gravity as a default.

mk56-spn commented:
Alright i think this pr has run its course, ill look into doing further adjustments as i play it more but thats it for now

bdach committed:
Remove unnecessary whitespace

bdach committed:
Reorder easing mapping to match enum order and throw on unknown animation style

bdach commented:
Have applied some of my changes on top of the existing code. I'm still unsure if we want this but at least the code seems sane at this point.

bdach reviewed (dismissed):
approving but not merging without a do-over

mk56-spn commented:
> approving but not merging without a do-over

I looked over them, why did you add a default? this isnt me questioning you its just i wanna know the reason for the purpose of not repeating the same mistake . is it like just considered good practice? since  i cant see how it would ever show up

bdach commented:
If you're asking about 5001412, then I prefer enum mappings to throw on unknown values rather than do some sort of fallback to an arbitrary value, which is what the `default` there does.

mk56-spn commented:
@peppy  dk about that title change, i left it vague because it does other things aswell, infact linear mode is just something i added after, its main purpose was to allow lower values , not that it really matters

peppy commented:
thanks for your opinion. pull request titles are at my discretion though. i use them for changelog.
